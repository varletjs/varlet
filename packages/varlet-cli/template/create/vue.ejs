<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
