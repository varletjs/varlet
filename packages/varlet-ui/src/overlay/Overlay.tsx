import { defineComponent, Teleport, Transition } from 'vue'
import { props } from './props'
import { useLock } from '../context/lock'
import { useZIndex } from '../context/zIndex'
import { createNamespace, useTeleport, call } from '../utils/components'

import '../styles/common.less'
import './overlay.less'

const { n, classes } = createNamespace('overlay')

export default defineComponent({
  name: 'VarOverlay',
  inheritAttrs: false,
  props,
  setup(props, { slots, attrs }) {
    const { zIndex } = useZIndex(() => props.show, 1)
    const { disabled } = useTeleport()
    const hideOverlay = () => {
      call(props.onClick)
      call(props['onUpdate:show'], false)
    }

    useLock(
      () => props.show,
      () => props.lockScroll
    )

    const renderOverlay = () => {
      return (
        <div
          class={classes(n())}
          style={{
            zIndex: zIndex.value - 1,
          }}
          {...attrs}
          onClick={hideOverlay}
        >
          {call(slots.default)}
        </div>
      )
    }

    const renderTransitionOverlay = () => {
      const { show } = props
      return <Transition name={n('$-fade')}>{show && renderOverlay()}</Transition>
    }
    return () => {
      const { teleport } = props
      if (teleport) {
        return (
          <Teleport to={teleport} disabled={disabled.value}>
            {renderTransitionOverlay()}
          </Teleport>
        )
      }
      return renderTransitionOverlay()
    }
  },
})
