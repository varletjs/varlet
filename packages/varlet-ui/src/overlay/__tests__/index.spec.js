import Overlay from '..'
import VarOverlay from '../Overlay'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'

test('test overlay plugin', () => {
  const app = createApp({}).use(Overlay)
  expect(app.component(Overlay.name)).toBeTruthy()
})

const Wrapper = {
  components: {
    [VarOverlay.name]: VarOverlay,
  },
  props: ['onClick'],
  data: () => ({
    show: false,
  }),
  template: `
    <var-overlay v-model:show="show" v-bind="$props">
      default slot content
    </var-overlay>
  `,
}

test('test overlay show', async () => {
  const wrapper = mount(Wrapper)
  expect(document.body.querySelector('.var-overlay')).toBeFalsy()

  await wrapper.setData({ show: true })
  expect(wrapper.find('.var-overlay').isVisible()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setData({ show: false })
  expect(document.body.querySelector('.var-overlay')).toBeFalsy()

  wrapper.unmount()
})

test('test overlay click on clickOverlay', async () => {
  const onClick = jest.fn()

  const wrapper = mount(Wrapper, {
    props: {
      onClick,
    },
  })

  await wrapper.setData({ show: true })

  await wrapper.find('.var-overlay').trigger('click')
  expect(onClick).toHaveBeenCalledTimes(1)

  await wrapper.setData({ show: true })
  await wrapper.setProps({ closeOnClickOverlay: false })
  await wrapper.find('.var-overlay').trigger('click')
  expect(onClick).toHaveBeenCalledTimes(2)

  wrapper.unmount()
})

test('test overlay z-index', async () => {
  const wrapper = mount(Wrapper)

  await wrapper.setData({ show: true })

  const prevOverlayZIndex = window.getComputedStyle(wrapper.find('.var-overlay').element).zIndex

  await wrapper.setData({ show: false })
  await wrapper.setData({ show: true })
  expect(window.getComputedStyle(wrapper.find('.var-overlay').element).zIndex).toBe(String(+prevOverlayZIndex + 1))

  wrapper.unmount()
})
