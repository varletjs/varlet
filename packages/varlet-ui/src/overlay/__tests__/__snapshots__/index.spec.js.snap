// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test overlay show 1`] = `"<div class=\\"var-overlay__overlay\\" style=\\"z-index: 2000;\\"> default slot content </div>"`;

exports[`test popup show 1`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 1998; display: none;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 1999;\\"></div>
    <transition-stub>
      <!---->
    </transition-stub>
  </div>
</transition-stub>"
`;

exports[`test popup show 2`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 2001;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
    <transition-stub>
      <div class=\\"var-popup__content var-elevation--3 var-popup--center\\" style=\\"z-index: 2003;\\"> default slot content </div>
    </transition-stub>
  </div>
</transition-stub>"
`;

exports[`test popup show 3`] = `
"<transition-stub>
  <div class=\\"var--box var-popup\\" style=\\"z-index: 2001; display: none;\\">
    <div class=\\"var-popup__overlay\\" style=\\"z-index: 2002;\\"></div>
    <transition-stub>
      <!---->
    </transition-stub>
  </div>
</transition-stub>"
`;
